dspace-api/src/main/java/org/dspace/authenticate/clarin/ShibGroup.java

```diff
@@ -0,0 +1,307 @@
+                    }
+
+                    // get the group ids
+                    groups.addAll(string2groups(groupNames));
+
+                } // foreach affiliations
+            } // if affiliations
+
+            //attribute -> group mapping
+            //check shibboleth attribute ATTR and put users having value ATTR_VALUE1 and ATTR_VALUE2 to GROUP1
+            //users having ATTR_VALUE3 to GROUP2
+            //groups must exist
+            //header.ATTR=ATTR_VALUE1=>GROUP1,ATTR_VALUE2=>GROUP1,ATTR_VALUE3=>GROUP2
+            final String lookFor = "authentication-shibboleth.header.";
+            ConfigurationService configurationService = new DSpace().getConfigurationService();
+            Properties allShibbolethProperties = configurationService.getProperties();
+            for (String propertyName : allShibbolethProperties.stringPropertyNames()) {
+                //look for properties in authentication shibboleth that start with "header."
+                if (propertyName.startsWith(lookFor)) {
+                    String headerName = propertyName.substring(lookFor.length());
+                    List<String> presentHeaderValues = shib_headers_.get(headerName);
+                    if (!CollectionUtils.isEmpty(presentHeaderValues)) {
+                        //if shibboleth sent any attributes under the headerName
+                        String[] values2groups = configurationService.getPropertyAsType(
+                                propertyName, String[].class);
+                        for (String value2group : values2groups) {
+                            String[] value2groupParts = value2group.split("=>", 2);
+                            String headerValue = value2groupParts[0].trim();
+                            String assignedGroup = value2groupParts[1].trim();
+                            if (presentHeaderValues.contains(headerValue)) {
+                                //our configured header value is present so add a group
+                                groups.addAll(string2groups(assignedGroup));
+                            }
+                        }
+                    }
+                }
+            }
+
+            /* <UFAL>
+             * Default group for shib authenticated users
+             */
+            Group default_group = get_default_group();
+            if ( null != default_group ) {
+                groups.add(default_group.getID());
+            }
+            /* </UFAL> */
+
+            log.info("Added current EPerson to special groups: " + groups);
+            // Convert from a Java Set to primitive ArrayList array
+            return new ArrayList<>(groups);
+        } catch (Throwable t) {
+            log.error(
+                    "Unable to validate any special groups this user may belong too because of an exception.",t);
+            return new ArrayList<>();
+        }
+    }
+
+    //
+    //
+    private List<String> get_affilations_from_roles(String roleHeader) {
```

Comment:
Was this our initiative to have the method names not camelCased?

Reply:
No, this code is copied from UFAL